curriculum/challenges/english/08-coding-interview-prep/project-euler/problem-35-circular-primes.english.md

```diff
@@ -63,37 +63,55 @@ circularPrimes(1000000);
 
 
 ```js
-const circularPrimes = (n) => {
-  const primeCheck = (num) => {
-    if (num === 1) {
-      return false;
-    }
-    for (let i = 2; i <= Math.floor(Math.sqrt(num)); i++) {
-      if (num % i === 0) {
-        return false;
-      }
+function rotate(n) {
+  if (String(n).length == 1) return n;
+  return "" + (n % 10) + Math.floor(n / 10);
+}
+function circularPrimes(n) {
+  const primes = [];
+  let count = 0;
+
+  // Making primes array
+  for (let i = 0; i < n; i++) {
+    primes.push(i);
+  }
+  primes[1] = 0;
+
+  // Getting upperbound
+  const upperBound = Math.ceil(Math.sqrt(n));
+
+  // Treating multiples of 2 as special cases
+  for (let i = 4; i < n; i += 2) {
+    primes[i] = 0;
+  }
+
+  // Setting other non-prime numbers to 0
+  for (let i = 3; i < upperBound; i += 2) {
+    for (let j = i * i; j < n; j += i) {
```

Comment:
`i` loops till the `sqrt(n)` so yes.
This is optimized version of [Sieve of Eratosthenes](https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes#Algorithm_and_variants) algorithm.

Basically, we are finding all the non prime numbers till `n` and marking them to `0`.

Reply:
for example, when i = 9, we do not need to run this loop.

Reply:
Ok, I guess we can add `if (primes[i])` before running the inner loop? Since for `i = 9`, `primes[9]` will be equal to `0` (already set by i = 3). That way inner loop wouldn't execute for numbers which are already factor of other `i`.